### ISSUE TITLE
Can't understand the hard negative mining at loss part, could you explain?

### ISSUE THREAD

**Issue by kzs569** [yijingru/ANCIS-Pytorch#7]
`        # hard negtive mining
        pos_mask = pos_mask.squeeze(2)  # torch.Size([2, 24448])
        p_conf_batch = p_conf.view(-1, self.num_classes)  # torch.Size([48896,2])
        temp = self.log_sum_exp(p_conf_batch) - p_conf_batch.gather(dim=1, index=t_conf.view(-1, 1))
        temp = temp.view(batch_size, -1)
        temp[pos_mask] = 0.
        _, temp_idx = temp.sort(1, descending=True)
        _, idx_rank = temp_idx.sort(1)
        num_neg = torch.clamp(self.neg_pos_ratio * num_pos, max=pos_mask.size(1) - 1).squeeze(2)
        neg_mask = idx_rank < num_neg.expand_as(idx_rank)`

When I want to replace the SSD module with RetinaNet, the way to calculate the diceloss, especially the classification loss make me some trouble. Could you explain how this part of code do negative mining at loss part, and do you have the same try on RetinaNet and have some advise?

Thank you for you code sharing!!

**Comment by yijingru** [comment]
Sorry for my late reply. If you are using RetinaNet, I suggest you directly use their focal loss because the imbalance problem between the positive and negative bounding boxes gets severe for RetinaNet compared to SSD. There's some focal loss PyTorch version on the GitHub, you can have a try. By the way, the hard negative mining is one way to solve the imbalance problem, the focal loss is supposed to have better or equal results.
